### ISSUE TITLE
Fields: leverage the React Context

### ISSUE THREAD

**Issue by ahmehri** [commercetools/ui-kit#2049]
**Is your feature request related to a problem? Please describe.**

ui-kit `Fields` components [use Formik](https://uikit.commercetools.com/?path=/story/philosophy-guide--forms).

> We use Formik to handle our form state

However, this usage is not efficient as it involves a lot of props drilling. Let's take `TextField` as an example, you can see in the following code that all Formik related props are being repeated explicitly, `onChange` -> `formik.handleChange`, etc.

```js
function MyComponent() {
  const formik = useFormikContext();

  <TextField
    name="sku"
    title="SKU"
    value={formik.values.sku}
    errors={formik.errors.sku}
    touched={formik.touched.sku}
    onBlur={formik.handleBlur}
    onChange={formik.handleChange}
  />;
}
```

This can be slightly improved by using `getFieldProps`, but it's still a chore to repeat each time a `Field` component is used.

```diff
function MyComponent() {
  const formik = useFormikContext();

  <TextField
-   name="sku"
    title="SKU"
-    value={formik.values.sku}
-    errors={formik.errors.sku}
-    touched={formik.touched.sku}
-    onBlur={formik.handleBlur}
-    onChange={formik.handleChange}
+    {...formik.getFieldProps('sku')}
  />;
}
```

**Describe the solution you'd like**

The solution is already provided by Formik, [Leveraging the React Context](https://formik.org/docs/tutorial#leveraging-react-context). The idea is to update ui-kit `Fields` components to benefit from this solution. This means updating `TextField` to something similar to the following (copied from the mentioned Formik docs page above):

```js
function TextField(props) {
  // useField() returns [formik.getFieldProps(), formik.getFieldMeta()]
  // which we can spread on <input>. We can use field meta to show an error
  // message if the field is invalid and it has been touched (i.e. visited)
  const [field, meta] = useField(props);
  return (
    <>
      <label htmlFor={props.id || props.name}>{label}</label>
      <input className="text-input" {...field} {...props} />
      {meta.touched && meta.error ? (
        <div className="error">{meta.error}</div>
      ) : null}
    </>
  );
};
```

This means that finally the usage of `TextField` will be simplified to the following:

```diff
function MyComponent() {
-  const formik = useFormikContext();

  <TextField
    name="sku"
    title="SKU"
-   {...formik.getFieldProps('sku')}
+   // any other non related Formik props
  />;
}
```

**Comment by emmenko** [comment]
This is already kind of possible. Example from an existing component:

```js
const FormApplicationNameField = () => {
  const { isSubmitting } = useFormikContext();
  const [field, meta] = useField('name');
  return (
    <TextField
      title={<FormattedMessage {...messages.labelName} />}
      isRequired={true}
      {...field}
      touched={meta.touched}
      errors={meta.error}
      isDisabled={isSubmitting}
      horizontalConstraint={13}
    />
  );
};
```

Is that what you were looking for?

**Comment by ahmehri** [comment]
> This is already kind of possible. Example from an existing component:
> 
> ```js
> const FormApplicationNameField = () => {
>   const { isSubmitting } = useFormikContext();
>   const [field, meta] = useField('name');
>   return (
>     <TextField
>       title={<FormattedMessage {...messages.labelName} />}
>       isRequired={true}
>       {...field}
>       touched={meta.touched}
>       errors={meta.error}
>       isDisabled={isSubmitting}
>       horizontalConstraint={13}
>     />
>   );
> };
> ```
> 
> Is that what you were looking for?

No, the idea is that ui-kit `Fields` components use internally `useField`, so that using them:

- Implies that they are used inside a Formik context (ensured by the consumer)
- The consumer needs to pass only the field's `name` as prop, nothing more, everything else related to Formik will be handled internally by the ui-kit component

In your example, the consumer does all of that which is different from the original idea. This is the desired usage of ui-kit `Fields` component:

<img width="825" alt="CleanShot 2021-12-22 at 13 50 14@2x" src="https://user-images.githubusercontent.com/3668245/147095442-9ecdd374-36f6-4a9b-9930-13d99f772623.png">

**Comment by emmenko** [comment]
I see.

Even though Formik is what we use to manage the form state in our Custom Applications, the field components technically can be used with any other form sate management.
If I remember correctly, this was done on purpose, to not bind the components to Formik.

However, what you would like to have are simply components that wrap the Formik bindings. For that, we don't necessarily need to reimplement the uikit components but to have some kind of wrappers.

The example I provided does essentially that.

Nevertheless, there are cases where you do actually want to control which props are passed from Formik to the field component. If we abstract this away, we don't have the flexibility anymore.

---

I would suggest first to try using more the `useField` hook and just spreading the `{...field}`. This should already help you with having to define less "proxy props".

**Comment by tdeekens** [comment]
> No, the idea is that ui-kit Fields components use internally useField, so that using them:

I am not sure we should bind our Design System that tightly to a form library. Also if we use that form library internally. Actually it was implemented (maybe not documented clearly enough) on purpose as it is. Reasons for me personally coming to mind:

1. Other consumers of ui-kit might want to use other form libraries than formik and I think have already
2. Any breaking change to formik could be a breaking change to the fields
3. Migrating to another form library would be quite hard with incompatible fields

I can totally see your point in it easing writing forms. However, maybe we instead don't have to build this into ui-kit but into a intermediate layer we can start fleshing out internally. It almost feels like a rare HoC use case these days.

**Comment by ahmehri** [comment]
@emmenko @tdeekens thanks for your input.

I originally implemented this [intermediate layer/solution](https://github.com/commercetools/merchant-center-frontend/blob/50dbe2ee3cec4a16c0d783a2b8b776cee8a58e2b/packages-application/application-products/components/text-field-formik.js#L1) and then thought that I was introducing a new layer of indirection and divergence from using ui-kit components.

But now that it's clear that our design system is not coupled to Formik and that the statement, in ui-kit docs, mentioning that we're using Formik is just information that we're using it in commercetools, I think the right way to go here is to restore the usage of those intermediate components.

**Comment by emmenko** [comment]
Ok thanks for clarifying this.

To be honest, I still don't see a huge value in having such wrapper components and I believe we can go with `useField` approach first.

**Comment by ahmehri** [comment]
> Ok thanks for clarifying this.
> 
> To be honest, I still don't see a huge value in having such wrapper components and I believe we can go with `useField` approach first.

@emmenko I see it makes sense. So what you're suggesting is a nice solution without introducing trade-offs when introducing new wrapper components. This will avoid us diverging from using ui-kit components, and at the same time allow us to benefit from leveraging the React context using `useField`.

In this case, I agree and would even consider your solution to be the right way to go. I already flagged the wrapped component that I created to not be used anymore and added a TODO to remove its usages.

Thanks again for your suggestion.
